File: app/settings/validator.py
# app/settings/validator.py
"""
Extensible settings validator that validates YAML sections against inferred schemas.

The validator works by:
1. Inferring the schema from the existing settings structure
2. Allowing custom validators to be registered for specific sections
3. Validating that new values match the expected structure and types
"""

import yaml
import logging
from dataclasses import dataclass, field
from typing import Any, Callable, Optional

LOGGER = logging.getLogger(__name__)


@dataclass
class ValidationError:
    """Represents a single validation error."""
    path: str  # Dot-separated path to the problematic field
    message: str
    value: Any = None

    def __str__(self) -> str:
        if self.path:
            return f"{self.path}: {self.message}"
        return self.message


@dataclass
class ValidationResult:
    """Result of a validation operation."""
    is_valid: bool
    errors: list[ValidationError] = field(default_factory=list)
    parsed_value: Any = None

    def __bool__(self) -> bool:
        return self.is_valid

    @property
    def error_messages(self) -> list[str]:
        return [str(e) for e in self.errors]

    @property
    def first_error(self) -> Optional[str]:
        return str(self.errors[0]) if self.errors else None


# Type alias for custom validator functions
# Validators receive (value, reference_schema) and return ValidationResult
CustomValidator = Callable[[Any, Any], ValidationResult]


class SettingsValidator:
    """
    Validates settings sections against inferred or custom schemas.

    The validator is extensible - you can register custom validators for
    specific section keys that need special handling.
    """

    # Class-level registry of custom validators
    _custom_validators: dict[str, CustomValidator] = {}

    # Known types that should be validated strictly
    STRICT_TYPES = (str, int, float, bool)

    @classmethod
    def register_validator(cls, section_key: str, validator: CustomValidator) -> None:
        """
        Register a custom validator for a specific settings section.

        Args:
            section_key: The top-level key this validator handles (e.g., 'models')
            validator: A callable that takes (value, reference) and returns ValidationResult
        """
        cls._custom_validators[section_key] = validator
        LOGGER.debug(f"Registered custom validator for section: {section_key}")

    @classmethod
    def validate_yaml_string(cls, yaml_string: str, reference: Any = None) -> ValidationResult:
        """
        Validate a YAML string can be parsed and optionally matches a reference structure.

        Args:
            yaml_string: The YAML content to validate
            reference: Optional reference value to validate structure against

        Returns:
            ValidationResult with parsed value if valid
        """
        # First, try to parse the YAML
        try:
            parsed = yaml.safe_load(yaml_string)
        except yaml.YAMLError as e:
            error_msg = str(e)
            # Extract line number if available
            if hasattr(e, 'problem_mark') and e.problem_mark:
                line = e.problem_mark.line + 1
                col = e.problem_mark.column + 1
                error_msg = f"Line {line}, column {col}: {e.problem}"
            return ValidationResult(
                is_valid=False,
                errors=[ValidationError(path="", message=f"Invalid YAML syntax: {error_msg}")]
            )

        # If no reference provided, just check it parses
        if reference is None:
            return ValidationResult(is_valid=True, parsed_value=parsed)

        # Validate structure against reference
        return cls.validate_structure(parsed, reference)

    @classmethod
    def validate_section(
        cls,
        section_key: str,
        yaml_string: str,
        reference_settings: dict
    ) -> ValidationResult:
        """
        Validate a specific settings section.

        Args:
            section_key: The section being validated (e.g., 'models', 'search')
            yaml_string: The YAML content for this section
            reference_settings: The full reference settings dict

        Returns:
            ValidationResult with the parsed section if valid
        """
        # First parse the YAML
        parse_result = cls.validate_yaml_string(yaml_string)
        if not parse_result.is_valid:
            return parse_result

        parsed = parse_result.parsed_value

        # Check for custom validator
        if section_key in cls._custom_validators:
            reference = reference_settings.get(section_key)
            return cls._custom_validators[section_key](parsed, reference)

        # Get reference for this section
        reference = reference_settings.get(section_key)

        # If no reference exists, allow any valid YAML
        if reference is None:
            return ValidationResult(is_valid=True, parsed_value=parsed)

        # Validate structure matches reference
        return cls.validate_structure(parsed, reference, section_key)

    @classmethod
    def validate_structure(
        cls,
        value: Any,
        reference: Any,
        path: str = ""
    ) -> ValidationResult:
        """
        Recursively validate that a value matches the structure of a reference.

        This is a permissive validator that:
        - Allows extra keys in dicts (for extensibility)
        - Validates types match for scalar values
        - Recursively validates nested structures

        Args:
            value: The value to validate
            reference: The reference value to match structure against
            path: Current path for error messages

        Returns:
            ValidationResult
        """
        errors: list[ValidationError] = []

        # Handle None values
        if value is None and reference is not None:
            # Allow None if reference was a dict/list (clearing a section)
            if isinstance(reference, (dict, list)):
                return ValidationResult(is_valid=True, parsed_value=value)
            errors.append(ValidationError(
                path=path,
                message=f"Expected {type(reference).__name__}, got null",
                value=value
            ))
            return ValidationResult(is_valid=False, errors=errors)

        # If reference is None, allow anything
        if reference is None:
            return ValidationResult(is_valid=True, parsed_value=value)

        # Validate dicts
        if isinstance(reference, dict):
            if not isinstance(value, dict):
                errors.append(ValidationError(
                    path=path,
                    message=f"Expected a mapping/object, got {type(value).__name__}",
                    value=value
                ))
                return ValidationResult(is_valid=False, errors=errors)

            # Recursively validate each key that exists in both
            for key in value:
                child_path = f"{path}.{key}" if path else key
                if key in reference:
                    child_result = cls.validate_structure(
                        value[key],
                        reference[key],
                        child_path
                    )
                    if not child_result.is_valid:
                        errors.extend(child_result.errors)
                # Extra keys are allowed (extensibility)

            return ValidationResult(
                is_valid=len(errors) == 0,
                errors=errors,
                parsed_value=value
            )

        # Validate lists
        if isinstance(reference, list):
            if not isinstance(value, list):
                errors.append(ValidationError(
                    path=path,
                    message=f"Expected a list, got {type(value).__name__}",
                    value=value
                ))
                return ValidationResult(is_valid=False, errors=errors)

            # If reference has items, validate each item against first reference item
            if reference and value:
                ref_item = reference[0]
                for i, item in enumerate(value):
                    child_path = f"{path}[{i}]"
                    child_result = cls.validate_structure(item, ref_item, child_path)
                    if not child_result.is_valid:
                        errors.extend(child_result.errors)

            return ValidationResult(
                is_valid=len(errors) == 0,
                errors=errors,
                parsed_value=value
            )

        # Validate scalar types
        if isinstance(reference, cls.STRICT_TYPES):
            # Allow type coercion for numbers
            if isinstance(reference, (int, float)) and isinstance(value, (int, float)):
                return ValidationResult(is_valid=True, parsed_value=value)

            # Otherwise check exact type
            if not isinstance(value, type(reference)):
                errors.append(ValidationError(
                    path=path,
                    message=f"Expected {type(reference).__name__}, got {type(value).__name__}",
                    value=value
                ))
                return ValidationResult(is_valid=False, errors=errors)

        return ValidationResult(is_valid=True, parsed_value=value)


# ============================================================================
# Built-in Custom Validators for Specific Sections
# ============================================================================

def _validate_model_config(model_config: Any, path: str, errors: list[ValidationError]) -> None:
    """Helper to validate a single model configuration entry."""
    if not isinstance(model_config, dict):
        errors.append(ValidationError(
            path=path,
            message="Model configuration must be a mapping"
        ))
        return

    # Check for required 'name' field
    if 'name' not in model_config:
        errors.append(ValidationError(
            path=path,
            message="Model configuration must have a 'name' field"
        ))

    # Validate context_length if present
    if 'context_length' in model_config:
        ctx = model_config['context_length']
        if not isinstance(ctx, int) or ctx <= 0:
            errors.append(ValidationError(
                path=f"{path}.context_length",
                message="context_length must be a positive integer"
            ))

    # Validate provider if present
    if 'provider' in model_config:
        provider = model_config['provider']
        valid_providers = ['ollama_chat', 'ollama', 'openai', 'anthropic', 'litellm']
        if provider not in valid_providers:
            errors.append(ValidationError(
                path=f"{path}.provider",
                message=f"Invalid provider. Must be one of: {', '.join(valid_providers)}"
            ))


def validate_models_section(value: Any, reference: Any) -> ValidationResult:
    """
    Custom validator for the 'models' section with model-specific rules.

    Structure:
        models:
          default:        # Default model config
            name: "..."
            provider: "..."
            context_length: 128000
          agent:          # Default for all agents
            name: "..."
          extractor:      # For memory extraction
            name: "..."
          agents:         # Per-agent overrides
            user_proxy_agent:
              name: "..."
            prime_agent:
              name: "..."
    """
    errors: list[ValidationError] = []

    if not isinstance(value, dict):
        return ValidationResult(
            is_valid=False,
            errors=[ValidationError("", "Models section must be a mapping")]
        )

    # Validate each top-level model entry
    for model_key, model_config in value.items():
        if model_key == 'agents':
            # Handle the nested agents subsection
            if not isinstance(model_config, dict):
                errors.append(ValidationError(
                    path="agents",
                    message="agents subsection must be a mapping"
                ))
                continue

            # Validate each agent's model config
            for agent_name, agent_model_config in model_config.items():
                _validate_model_config(
                    agent_model_config,
                    f"agents.{agent_name}",
                    errors
                )
        else:
            # Regular model config (default, agent, extractor, etc.)
            _validate_model_config(model_config, model_key, errors)

    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors,
        parsed_value=value
    )


def validate_search_section(value: Any, reference: Any) -> ValidationResult:
    """Custom validator for the 'search' section."""
    errors: list[ValidationError] = []

    if not isinstance(value, dict):
        return ValidationResult(
            is_valid=False,
            errors=[ValidationError("", "Search section must be a mapping")]
        )

    # Validate provider
    if 'provider' in value:
        valid_providers = ['brave', 'google', 'duckduckgo', 'serper']
        if value['provider'] not in valid_providers:
            errors.append(ValidationError(
                path="provider",
                message=f"Invalid search provider. Must be one of: {', '.join(valid_providers)}"
            ))

    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors,
        parsed_value=value
    )


def validate_mcp_servers_section(value: Any, reference: Any) -> ValidationResult:
    """Custom validator for the 'mcp_servers' section."""
    errors: list[ValidationError] = []

    if not isinstance(value, dict):
        return ValidationResult(
            is_valid=False,
            errors=[ValidationError("", "MCP servers section must be a mapping")]
        )

    for server_name, server_config in value.items():
        if not isinstance(server_config, dict):
            errors.append(ValidationError(
                path=server_name,
                message="MCP server configuration must be a mapping"
            ))
            continue

        # Check server type
        server_type = server_config.get('type', 'stdio')

        if server_type == 'http':
            # HTTP servers need a url
            if 'url' not in server_config:
                errors.append(ValidationError(
                    path=server_name,
                    message="HTTP MCP server must have a 'url' field"
                ))
        else:
            # stdio servers need a command
            if 'command' not in server_config:
                errors.append(ValidationError(
                    path=server_name,
                    message="MCP server must have a 'command' field"
                ))

    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors,
        parsed_value=value
    )


def validate_agent_prompts_section(value: Any, reference: Any) -> ValidationResult:
    """Custom validator for the 'agent_prompts' section."""
    errors: list[ValidationError] = []

    if not isinstance(value, dict):
        return ValidationResult(
            is_valid=False,
            errors=[ValidationError("", "Agent prompts section must be a mapping")]
        )

    for agent_name, prompt in value.items():
        if not isinstance(prompt, str):
            errors.append(ValidationError(
                path=agent_name,
                message="Agent prompt must be a string"
            ))
            continue

        if len(prompt.strip()) < 10:
            errors.append(ValidationError(
                path=agent_name,
                message="Agent prompt appears too short (less than 10 characters)"
            ))

    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors,
        parsed_value=value
    )


# Register built-in validators
SettingsValidator.register_validator('models', validate_models_section)
SettingsValidator.register_validator('search', validate_search_section)
SettingsValidator.register_validator('mcp_servers', validate_mcp_servers_section)
SettingsValidator.register_validator('agent_prompts', validate_agent_prompts_section)

File: solenoid-node/src/config/validator.ts
/**
 * Extensible settings validator that validates YAML sections against inferred schemas.
 *
 * The validator works by:
 * 1. Using Zod schemas for base validation
 * 2. Allowing custom validators to be registered for specific sections
 * 3. Providing detailed error messages with paths
 */

import { z, type ZodError } from 'zod';
import { parse as parseYaml, YAMLParseError } from 'yaml';
import {
  AppSettingsSchema,
  ModelsConfigSchema,
  SearchConfigSchema,
  McpServerSchema,
  AgentPromptsSchema,
  EmbeddingsConfigSchema,
  type AppSettings,
} from './schema.js';

export interface ValidationError {
  path: string;
  message: string;
  value?: unknown;
}

export interface ValidationResult {
  isValid: boolean;
  errors: ValidationError[];
  parsedValue?: unknown;
}

export type CustomValidator = (value: unknown, reference?: unknown) => ValidationResult;

// Section key type
export type SectionKey = keyof AppSettings;

// Section info for UI display
export interface SectionInfo {
  key: SectionKey;
  displayName: string;
  description: string;
}

// Section metadata for the UI
export const SECTION_INFO: Record<SectionKey, SectionInfo> = {
  models: {
    key: 'models',
    displayName: 'Models',
    description: 'Configure model settings (defaults and per-agent overrides)',
  },
  search: {
    key: 'search',
    displayName: 'Search',
    description: 'Configure web search provider and API keys',
  },
  mcp_servers: {
    key: 'mcp_servers',
    displayName: 'MCP Servers',
    description: 'Configure Model Context Protocol server connections',
  },
  agent_prompts: {
    key: 'agent_prompts',
    displayName: 'Agent Prompts',
    description: 'Configure system prompts for each agent',
  },
  embeddings: {
    key: 'embeddings',
    displayName: 'Embeddings',
    description: 'Configure embedding model settings',
  },
};

// Map section keys to their Zod schemas
const SECTION_SCHEMAS: Record<SectionKey, z.ZodType<unknown>> = {
  models: ModelsConfigSchema,
  search: SearchConfigSchema,
  mcp_servers: z.record(z.string(), McpServerSchema),
  agent_prompts: AgentPromptsSchema,
  embeddings: EmbeddingsConfigSchema,
};

// Custom validators registry
const customValidators: Map<SectionKey, CustomValidator> = new Map();

/**
 * Register a custom validator for a specific settings section.
 */
export function registerValidator(sectionKey: SectionKey, validator: CustomValidator): void {
  customValidators.set(sectionKey, validator);
}

/**
 * Convert Zod errors to our ValidationError format.
 */
function zodErrorsToValidationErrors(zodError: ZodError): ValidationError[] {
  return zodError.errors.map((err) => ({
    path: err.path.join('.'),
    message: err.message,
  }));
}

/**
 * Validate a YAML string can be parsed.
 */
export function validateYamlString(yamlString: string): ValidationResult {
  try {
    const parsed = parseYaml(yamlString);
    return {
      isValid: true,
      errors: [],
      parsedValue: parsed,
    };
  } catch (error) {
    if (error instanceof YAMLParseError) {
      const lineInfo = error.linePos
        ? `Line ${error.linePos[0].line}, column ${error.linePos[0].col}: `
        : '';
      return {
        isValid: false,
        errors: [
          {
            path: '',
            message: `Invalid YAML syntax: ${lineInfo}${error.message}`,
          },
        ],
      };
    }
    return {
      isValid: false,
      errors: [
        {
          path: '',
          message: `Invalid YAML: ${error instanceof Error ? error.message : 'Unknown error'}`,
        },
      ],
    };
  }
}

/**
 * Validate a specific settings section.
 */
export function validateSection(
  sectionKey: SectionKey,
  yamlString: string,
  _referenceSettings?: AppSettings
): ValidationResult {
  // First parse the YAML
  const parseResult = validateYamlString(yamlString);
  if (!parseResult.isValid) {
    return parseResult;
  }

  const parsed = parseResult.parsedValue;

  // Check for custom validator first
  const customValidator = customValidators.get(sectionKey);
  if (customValidator) {
    return customValidator(parsed, _referenceSettings?.[sectionKey]);
  }

  // Use Zod schema validation
  const schema = SECTION_SCHEMAS[sectionKey];
  if (!schema) {
    return {
      isValid: true,
      errors: [],
      parsedValue: parsed,
    };
  }

  const result = schema.safeParse(parsed);
  if (!result.success) {
    return {
      isValid: false,
      errors: zodErrorsToValidationErrors(result.error),
    };
  }

  return {
    isValid: true,
    errors: [],
    parsedValue: result.data,
  };
}

/**
 * Validate the entire settings object.
 */
export function validateSettings(yamlString: string): ValidationResult {
  const parseResult = validateYamlString(yamlString);
  if (!parseResult.isValid) {
    return parseResult;
  }

  const result = AppSettingsSchema.safeParse(parseResult.parsedValue);
  if (!result.success) {
    return {
      isValid: false,
      errors: zodErrorsToValidationErrors(result.error),
    };
  }

  return {
    isValid: true,
    errors: [],
    parsedValue: result.data,
  };
}

// ============================================================================
// Built-in Custom Validators for Specific Sections
// ============================================================================

/**
 * Custom validator for the 'models' section with model-specific rules.
 */
function validateModelsSection(value: unknown, _reference?: unknown): ValidationResult {
  const errors: ValidationError[] = [];

  if (!value || typeof value !== 'object') {
    return {
      isValid: false,
      errors: [{ path: '', message: 'Models section must be an object' }],
    };
  }

  const models = value as Record<string, unknown>;

  // Validate each model entry
  for (const [modelKey, modelConfig] of Object.entries(models)) {
    if (modelKey === 'agents') {
      // Handle the nested agents subsection
      if (modelConfig && typeof modelConfig !== 'object') {
        errors.push({
          path: 'agents',
          message: 'agents subsection must be an object',
        });
        continue;
      }

      if (modelConfig) {
        for (const [agentName, agentModelConfig] of Object.entries(
          modelConfig as Record<string, unknown>
        )) {
          validateModelConfig(agentModelConfig, `agents.${agentName}`, errors);
        }
      }
    } else {
      // Regular model config (default, agent, extractor, etc.)
      validateModelConfig(modelConfig, modelKey, errors);
    }
  }

  // Also validate with Zod for complete validation
  const zodResult = ModelsConfigSchema.safeParse(value);
  if (!zodResult.success) {
    errors.push(...zodErrorsToValidationErrors(zodResult.error));
  }

  return {
    isValid: errors.length === 0,
    errors,
    parsedValue: value,
  };
}

function validateModelConfig(
  config: unknown,
  path: string,
  errors: ValidationError[]
): void {
  if (!config || typeof config !== 'object') {
    errors.push({
      path,
      message: 'Model configuration must be an object',
    });
    return;
  }

  const modelConfig = config as Record<string, unknown>;

  // Check for required 'name' field (except for partial configs in agents)
  if (path === 'default' && !modelConfig.name) {
    errors.push({
      path,
      message: "Model configuration must have a 'name' field",
    });
  }

  // Validate context_length if present
  if ('context_length' in modelConfig) {
    const ctx = modelConfig.context_length;
    if (typeof ctx !== 'number' || ctx <= 0 || !Number.isInteger(ctx)) {
      errors.push({
        path: `${path}.context_length`,
        message: 'context_length must be a positive integer',
      });
    }
  }

  // Validate provider if present
  if ('provider' in modelConfig) {
    const provider = modelConfig.provider;
    const validProviders = ['ollama_chat', 'ollama', 'openai', 'anthropic', 'litellm'];
    if (typeof provider !== 'string' || !validProviders.includes(provider)) {
      errors.push({
        path: `${path}.provider`,
        message: `Invalid provider. Must be one of: ${validProviders.join(', ')}`,
      });
    }
  }
}

/**
 * Custom validator for the 'search' section.
 */
function validateSearchSection(value: unknown, _reference?: unknown): ValidationResult {
  const errors: ValidationError[] = [];

  if (!value || typeof value !== 'object') {
    return {
      isValid: false,
      errors: [{ path: '', message: 'Search section must be an object' }],
    };
  }

  const search = value as Record<string, unknown>;

  // Validate provider
  if ('provider' in search) {
    const validProviders = ['brave', 'google', 'duckduckgo', 'serper', 'none'];
    if (typeof search.provider !== 'string' || !validProviders.includes(search.provider)) {
      errors.push({
        path: 'provider',
        message: `Invalid search provider. Must be one of: ${validProviders.join(', ')}`,
      });
    }
  }

  return {
    isValid: errors.length === 0,
    errors,
    parsedValue: value,
  };
}

/**
 * Custom validator for the 'mcp_servers' section.
 */
function validateMcpServersSection(value: unknown, _reference?: unknown): ValidationResult {
  const errors: ValidationError[] = [];

  if (!value || typeof value !== 'object') {
    return {
      isValid: false,
      errors: [{ path: '', message: 'MCP servers section must be an object' }],
    };
  }

  const servers = value as Record<string, unknown>;

  for (const [serverName, serverConfig] of Object.entries(servers)) {
    if (!serverConfig || typeof serverConfig !== 'object') {
      errors.push({
        path: serverName,
        message: 'MCP server configuration must be an object',
      });
      continue;
    }

    const config = serverConfig as Record<string, unknown>;
    const serverType = config.type ?? 'stdio';

    if (serverType === 'http') {
      // HTTP servers need a url
      if (!config.url) {
        errors.push({
          path: serverName,
          message: "HTTP MCP server must have a 'url' field",
        });
      }
    } else {
      // stdio servers need a command
      if (!config.command) {
        errors.push({
          path: serverName,
          message: "MCP server must have a 'command' field",
        });
      }
    }
  }

  return {
    isValid: errors.length === 0,
    errors,
    parsedValue: value,
  };
}

/**
 * Custom validator for the 'agent_prompts' section.
 */
function validateAgentPromptsSection(value: unknown, _reference?: unknown): ValidationResult {
  const errors: ValidationError[] = [];

  if (!value || typeof value !== 'object') {
    return {
      isValid: false,
      errors: [{ path: '', message: 'Agent prompts section must be an object' }],
    };
  }

  const prompts = value as Record<string, unknown>;

  for (const [agentName, prompt] of Object.entries(prompts)) {
    if (typeof prompt !== 'string') {
      errors.push({
        path: agentName,
        message: 'Agent prompt must be a string',
      });
      continue;
    }

    if (prompt.trim().length < 10) {
      errors.push({
        path: agentName,
        message: 'Agent prompt appears too short (less than 10 characters)',
      });
    }
  }

  return {
    isValid: errors.length === 0,
    errors,
    parsedValue: value,
  };
}

// Register built-in validators
registerValidator('models', validateModelsSection);
registerValidator('search', validateSearchSection);
registerValidator('mcp_servers', validateMcpServersSection);
registerValidator('agent_prompts', validateAgentPromptsSection);
